doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,968 @@
+
+
+    class MyPass : public BasicBlockPass {
+      std::vector<Value *> IntegerVect;
+
+    public:
+
+      static char ID;
+
+      MyPass() : BasicBlockPass(ID) {}
+
+      virtual bool runOnBasicBlock(BasicBlock &BB) override {
+        IntegerVect.clear();
+
+        // Not iterating from the beginning to avoid obfuscation of Phi instructions
+        // parameters
+        for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
+                                           end = BB.end();
+             I != end; ++I) {
+          Instruction &Inst = *I;
+            for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
+              if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
+                errs() << "I've found one sir!\n";
+              }
+          }
+          registerInteger(Inst);
+        }
+        return false;
+      }
+
+    private:
+      void registerInteger(Value &V) {
+        if (V.getType()->isIntegerTy()) {
+          IntegerVect.emplace_back(&V);
+          errs() << "Registering an integer!" << V << "\n";
+        }
+      }
+
+      Constant *isValidCandidateOperand(Value *V) {
+        if (Constant *C = dyn_cast<Constant>(V)) {
+          // Checking constant eligibility
+          if (isa<PointerType>(C->getType())) {
+            // dbgs() << "Ignoring NULL pointers\n";
+            return nullptr;
+          } else if (C->getType()->isFloatingPointTy()) {
+            // dbgs() << "Ignoring Floats 0\n";
+            return nullptr;
+          } else if (C->isNullValue()) {
+            return C;
+          } else {
+            return nullptr;
+          }
+        } else {
+          return nullptr;
+        }
+      }
+    };
+
+
+and replace your test code by this updated version:
+
+.. code:: c
+
+    #include <stdio.h>
+
+    int foo(){return 1;}
+
+    int main() {
+        int a = 2;
+        printf("Hello world\n");
+        a *= 3;
+
+        return 0;
+    }
+
+
+If you run the pass on our new test file you'll notice that the pass finds **3** integers to register corresponding to %2, %3 and %4 in the following bytecode:
+
+.. code:: llvm
+
+    ; Function Attrs: nounwind uwtable
+    define i32 @main() #0 {
+      %1 = alloca i32, align 4
+      %a = alloca i32, align 4
+      store i32 0, i32* %1
+      store i32 2, i32* %a, align 4
+      %2 = call i32 @puts(i8* getelementptr inbounds ([13 x i8]* @.str, i32 0, i32 0))
+      %3 = load i32* %a, align 4
+      %4 = mul nsw i32 %3, 3
+      store i32 %4, i32* %a, align 4
+      ret i32 0
+    }
+
+
+There are a few things that you should remember from this little modification:
+* The LLVM bytecode is in `SSA form <http://en.wikipedia.org/wiki/Static_single_assignment_form>`_, so you will see variables that you didn't explicitly declared appear in the bytecode. Typically temporary result or ``load``s.
+* A variable declaration in your code returns a **pointer** in the bytecode not an instance of the type of the variable.
```

Comment:
Clang translates variable declaration into variable allocated on the stack (through the alloca instruction ). A later pass (Mem2reg) takes care of putting them in registers when possible.